```
checking host system type... Invalid configuration `i686-w64-mingw32-gcc': machine `i686-w64-mingw32' not recognized
configure: error: /bin/sh ./config.sub i686-w64-mingw32-gcc failed
make: *** [Makefile:72: /cygdrive/c/projects/fftw-builder/build/lib/libfftw3f-3.dll] Error 1
make: Leaving directory '/cygdrive/c/projects/fftw-builder'
```
Worked passing `CC` directly though. Any idea why?

--- a/.travis.yml
+++ b/.travis.yml
@@ -6,11 +6,16 @@ services:
 matrix:
   include:
     - os: linux
-      env: ARCH="i686" DOCKER_IMAGE="staticfloat/julia_workerbase:centos6_9-x86"
+      env: ARCH="i686" DOCKER_IMAGE="staticfloat/julia_workerbase:centos6_9-x86" JULIA_URL="https://julialangnightlies-s3.julialang.org/bin/linux/x86/julia-latest-linux32.tar.gz"
+      addons:
+        apt:
+          packages:
+            - libc6:i386
+            - zlib1g:i386
     - os: linux
-      env: ARCH="x86_64" DOCKER_IMAGE="staticfloat/julia_workerbase:centos6_9-x64"
+      env: ARCH="x86_64" DOCKER_IMAGE="staticfloat/julia_workerbase:centos6_9-x64" JULIA_URL="https://julialangnightlies-s3.julialang.org/bin/linux/x64/julia-latest-linux64.tar.gz"
     - os: osx
-      env: ARCH="x86_64"
+      env: ARCH="x86_64" JULIA_URL="https://julialangnightlies-s3.julialang.org/bin/osx/x64/julia-latest-osx.dmg"
       osx_image: xcode8
 branches:
   only:
@@ -21,13 +26,26 @@ before_script:
   - if [ "$(uname)" = "Linux" ]; then
       docker pull $DOCKER_IMAGE;
     fi
+  - export CURL_USER_AGENT="Travis-CI $(curl --version | head -n 1)"
+  - if [ "$(uname)" = "Linux" ]; then
+      mkdir -p ~/julia;
+      curl -A "${CURL_USER_AGENT}" -s -L --retry 7 "${JULIA_URL}" | tar -C ~/julia -x -z --strip-components=1 -f -;
+    fi
+  - if [ "$(uname)" = "Darwin" ]; then
+      curl -A "${CURL_USER_AGENT}" -s -L --retry 7 -o julia.dmg "${JULIA_URL}";
+      mkdir juliamnt;
+      hdiutil mount -readonly -mountpoint juliamnt julia.dmg;
+      cp -a juliamnt/*.app/Contents/Resources/julia ~/;
+    fi
+  - export PATH="${PATH}:${HOME}/julia/bin"
+  - julia -e 'versioninfo()'
 script:
   - if [ "$(uname)" = "Linux" ]; then
       docker run --user=$(id -u):$(id -g) -ti -v $(pwd):/src -w /src $DOCKER_IMAGE make;
     else
       make;
     fi
-  - if [ -z "$(ls build/lib/libfftw*)" ]; then exit 1; fi
+  - julia test.jl
 deploy:
   provider: releases
   api_key:

--- a/Makefile
+++ b/Makefile
@@ -3,7 +3,22 @@ SRCDIR := $(BUILDDIR)/src
 LIBDIR := $(BUILDDIR)/lib
 BINDIR := $(BUILDDIR)/bin
 
-default: $(LIBDIR)/libfftw%
+ifeq ($(OS),Windows_NT)
+SHLIB_EXT := dll
+SUFFIX := -3
+else
+SUFFIX := _threads
+ifeq ($(shell uname),Darwin)
+SHLIB_EXT := dylib
+else
+SHLIB_EXT := so
+endif
+endif
+
+TARGET_single := libfftw3f$(SUFFIX).$(SHLIB_EXT)
+TARGET_double := libfftw3$(SUFFIX).$(SHLIB_EXT)
+
+default: $(LIBDIR)/$(TARGET_single) $(LIBDIR)/$(TARGET_double)
 
 # Define directory-creation rule for many directories we'll need
 define dir_rule
@@ -23,14 +38,16 @@ ifeq ($(ARCH),x86_64)
 FFTW_CONFIG += --enable-sse2 --enable-fma
 endif
 
-ifeq ($(OS),WINNT)
+ifeq ($(OS),Windows_NT)
+CONFIG += --host=$(ARCH)-w64-mingw32
 FFTW_CONFIG += --with-our-malloc --with-combined-threads
 ifneq ($(ARCH),x86_64)
 FFTW_CONFIG += --with-incoming-stack-boundary=2
 endif
 endif
 
-FFTW_CONFIG += --enable-single
+FFTW_ENABLE_single := --enable-single
+FFTW_ENABLE_double :=
 
 .PHONY: default
 
@@ -40,10 +57,16 @@ $(SRCDIR)/fftw-$(FFTW_VERS).tar.gz: | $(SRCDIR)
 $(SRCDIR)/configure: $(SRCDIR)/fftw-$(FFTW_VERS).tar.gz
 	$(TAR) -C $(dir $@) --strip-components 1 -xf $<
 
-$(LIBDIR)/libfftw%: $(SRCDIR)/configure | $(LIBDIR) $(BINDIR)
-	# Try to configure with AVX support, if that fails then try again without
-	(cd $(dir $<) && \
-	    ./configure $(CONFIG) $(FFTW_CONFIG) --enable-avx || \
-	    ./configure $(CONFIG) $(FFTW_CONFIG))
-	$(MAKE) -C $(dir $<)
-	$(MAKE) -C $(dir $<) install
+define FFTW_BUILD
+$$(LIBDIR)/$$(TARGET_$1): $$(SRCDIR)/configure | $$(LIBDIR) $$(BINDIR)
+	(cd $$(dir $$<) && \
+	    ./configure $$(CONFIG) $$(FFTW_CONFIG) $$(FFTW_ENABLE_$1) --enable-avx || \
+	    ./configure $$(CONFIG) $$(FFTW_CONFIG) $$(FFTW_ENABLE_$1))
+	$(MAKE) -C $$(dir $$<)
+	$(MAKE) -C $$(dir $$<) install
+ifeq ($(OS),Windows_NT)
+	mv "$$(BINDIR)/$$(TARGET_$1)" "$$(LIBDIR)/$$(TARGET_$1)"
+endif
+endef
+
+$(foreach prec,single double,$(eval $(call FFTW_BUILD,$(prec))))

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -1,20 +1,40 @@
 environment:
   matrix:
    - ARCH: "i686"
+     JULIA_URL: "https://julialangnightlies-s3.julialang.org/bin/winnt/x86/julia-latest-win32.exe"
    - ARCH: "x86_64"
+     JULIA_URL: "https://julialangnightlies-s3.julialang.org/bin/winnt/x64/julia-latest-win64.exe"
 branches:
   only:
     - master
+skip_commits:
+  message: /\[av skip\]/
 notifications:
   - provider: Email
     on_build_success: false
     on_build_failure: false
     on_build_status_changed: false
-clone_depth: 1
+shallow_clone: true
+clone_depth: 5
 init:
   - git config --global core.autocrlf input
+  - ps: "[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12"
+install:
+  - appveyor DownloadFile https://cygwin.com/setup-x86_64.exe -FileName C:\projects\cygwin-setup-x86_64.exe
+  - C:\projects\cygwin-setup-x86_64.exe -q -g -P make,mingw64-i686-gcc-core,mingw64-x86_64-gcc-core
+  - appveyor DownloadFile %JULIA_URL% -FileName C:\projects\julia-binary.exe
+  - set PATH="C:\Program Files\7-Zip"
 build_script:
-  - C:\MinGW\bin\mingw32-make.exe
+  - C:\cygwin64\bin\sh -lc "make -C $(cygpath $APPVEYOR_BUILD_FOLDER)"
+# after_build:
+#   - 7z a <built binaries>
+before_test:
+  - C:\projects\julia-binary.exe /S /D=C:\projects\julia
+  - C:\projects\julia\bin\julia -e "versioninfo()"
+test_script:
+  - C:\projects\julia\bin\julia test.jl
+# artifacts:
+#   - path: <the stuff>
 
 # FROM AV DOCS:
 # deploy:

--- a/test.jl
+++ b/test.jl
@@ -1,0 +1,34 @@
+using Base.Test
+
+if Sys.iswindows()
+    builddir = joinpath(ENV["APPVEYOR_BUILD_FOLDER"], "build")
+    suffix = "-3." * Libdl.dlext
+else
+    builddir = joinpath(ENV["TRAVIS_BUILD_DIR"], "build")
+    suffix = "_threads." * Libdl.dlext
+end
+
+bindir = joinpath(builddir, "bin")
+libdir = joinpath(builddir, "lib")
+
+const libfftw = joinpath(libdir, "libfftw3$suffix")
+const libfftwf = joinpath(libdir, "libfftw3f$suffix")
+
+# For debugging
+println("=====\nFull contents of build/lib:")
+foreach(println, readdir(libdir))
+println("=====\nFull contents of build/bin:")
+foreach(println, readdir(bindir))
+println("=====\nLooking for:\n", libfftw, "\n", libfftwf)
+
+@testset "things exist" begin
+    @test isfile(libfftw)
+    @test isfile(libfftwf)
+end
+
+@testset "things work" begin
+    stat = ccall((:fftw_init_threads, libfftw), Int32, ())
+    @test stat > 0
+    plan2 = ccall((:fftw_plan_with_nthreads, libfftw), Void, (Int32,), Int32(2))
+    @test plan2 === nothing
+end

